File: Butterflies/JavaScript/Resources/Components/Spawner.js
'atomic component';

var viewport = Atomic.renderer.getViewport(0);

//cache resource
var particleEffect = Atomic.cache.getResource("ParticleEffect2D", "Particles/particle.pex");

exports.component = function(self) {

    self.start = function() {
        //touch control
        self.subscribeToEvent("TouchMove", function(event) {
            createButterflyNode([event.x, event.y]);
        });
        self.subscribeToEvent("MultiGesture", function(event) {
            if(event.numFingers >= 2) {
                createButterflyParticle([event.centerX, event.centerY]);
            }
        });
    };

    self.update = function(timeStep) {
        //if Left mouse button is pressed
        if (Atomic.input.getMouseButtonDown(Atomic.MOUSEB_LEFT)) {
            var mousePos = Atomic.input.getMousePosition();
            createButterflyNode(mousePos);
        //if Right mouse button WAS pressed once
        } else if (Atomic.input.getMouseButtonPress(Atomic.MOUSEB_RIGHT)) {
            var mousePos = Atomic.input.getMousePosition();
            createButterflyParticle(mousePos);
        }
    };

    function createButterflyNode(pos) {
      //project mouse screen position to the world position
      pos = viewport.screenToWorldPoint(pos[0], pos[1], 0);
      //create butterfly node
      var butterfly = self.scene.createChild("Butterfly");
      butterfly.position2D = pos;
      butterfly.createJSComponent("Components/Butterfly.js");
    }

    function createButterflyParticle(pos) {
      //create particle emitter
      var emitter = self.scene.createChild("ButterflyEmitter");
      //project mouse screen position to the world position
      pos = viewport.screenToWorldPoint(pos[0], pos[1], 0);
      emitter.position2D = pos;
      var pex = emitter.createComponent("ParticleEmitter2D");
      pex.effect = particleEffect;
    }
};

File: Butterflies/CSharp/Resources/Components/Spawner.cs
using System;
using AtomicEngine;

public class Spawner : CSComponent
{

    void Start()
    {
        viewport = AtomicNET.GetSubsystem<Renderer>().GetViewport(0);
    }

    void Update(float timeStep)
    {

        var input = AtomicNET.GetSubsystem<Input>();

        if (input.GetMouseButtonDown(Constants.MOUSEB_LEFT))
        {
            var mousePos = input.GetMousePosition();

            for (var i = 0; i < 10; i++)
                createButterflyNode(new Vector2(mousePos.X, mousePos.Y));
        }        
		else if (input.GetMouseButtonDown(Constants.MOUSEB_RIGHT))
		{
			var mousePos = input.GetMousePosition();

			createButterflyParticle(new Vector2(mousePos.X, mousePos.Y));
		}        

    }

    void createButterflyNode(Vector2 pos)
    {
        //project mouse screen position to the world position
        var screenPos = viewport.ScreenToWorldPoint((int) pos.X, (int) pos.Y, 0);

        //create butterfly node
        var node = Scene.CreateChild("Butterfly");

        node.Position2D = new Vector2(screenPos.X, screenPos.Y);

        var butterfly = node.CreateComponent<Butterfly>();

    }

	void createButterflyParticle(Vector2 pos)
	{
		//project mouse screen position to the world position
		var screenPos = viewport.ScreenToWorldPoint((int) pos.X, (int) pos.Y, 0);

		//create particle emitter
		var emitter = Scene.CreateChild("ButterflyEmitter");

		emitter.Position2D = new Vector2(screenPos.X, screenPos.Y);

		var pex = emitter.CreateComponent<ParticleEmitter2D>();

		pex.SetEffect(GetSubsystem<ResourceCache>().GetResource<ParticleEffect2D> ("Particles/particle.pex"));

	}
    Viewport viewport;


}
